File: apps/android/app/src/main/java/com/litter/android/state/AppLifecycleController.kt
package com.litter.android.state

import android.content.Context
import com.litter.android.ui.ExperimentalFeatures
import com.litter.android.util.LLog
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Mutex
import uniffi.codex_mobile_client.AppRefreshAccountRequest
import uniffi.codex_mobile_client.ThreadKey
import uniffi.codex_mobile_client.AppServerHealth

/**
 * Handles app lifecycle events: server reconnection on resume,
 * background turn tracking on pause, and push notification handling.
 */
class AppLifecycleController {
    companion object {
        internal fun reconnectCandidates(
            savedServers: List<SavedServer>,
            activeServerIds: Set<String>,
        ): List<SavedServer> =
            savedServers.filter { server ->
                server.source != "local" && server.id !in activeServerIds
            }
    }

    private val reconnectMutex = Mutex()

    /** Threads that were active when the app went to background. */
    private val backgroundedTurnKeys = mutableSetOf<ThreadKey>()

    /** FCM device push token. */
    var devicePushToken: String? = null
        private set

    fun setDevicePushToken(token: String) {
        devicePushToken = token
    }

    /**
     * Reconnects all saved servers on app launch or resume.
     */
    suspend fun reconnectSavedServers(context: Context, appModel: AppModel) {
        if (!reconnectMutex.tryLock()) {
            LLog.t("AppLifecycleController", "reconnect already in progress; skipping")
            return
        }
        try {
            ExperimentalFeatures.initialize(context.applicationContext)
            val sshCredentials = SshCredentialStore(context)
            val activeServerIds = appModel.store.snapshot()
                .servers
                .filter { it.health != AppServerHealth.DISCONNECTED }
                .mapTo(mutableSetOf()) { it.serverId }
            val saved = reconnectCandidates(
                savedServers = SavedServerStore.remembered(context),
                activeServerIds = activeServerIds,
            )
            coroutineScope {
                saved.map { server ->
                    async {
                        try {
                            reconnectSavedServer(appModel, server, sshCredentials)
                        } catch (e: Exception) {
                            // Best-effort reconnection — server may be offline
                            LLog.e(
                                "AppLifecycleController",
                                "saved server reconnect failed",
                                e,
                                fields = mapOf("serverId" to server.id, "host" to server.hostname, "os" to server.os),
                            )
                        }
                    }
                }.awaitAll()
            }
            appModel.refreshSnapshot()
        } finally {
            reconnectMutex.unlock()
        }
    }

    suspend fun reconnectServer(context: Context, appModel: AppModel, serverId: String) {
        val currentServer = appModel.store.snapshot().servers.firstOrNull { it.serverId == serverId }
        if (currentServer?.isLocal == true || serverId == "local") {
            appModel.restartLocalServer()
            return
        }

        val sshCredentials = SshCredentialStore(context)
        val savedServer = SavedServerStore.load(context).firstOrNull { it.id == serverId }
        if (savedServer != null) {
            appModel.sshSessionStore.close(serverId)
            runCatching { appModel.serverBridge.disconnectServer(serverId) }
            try {
                reconnectSavedServer(appModel, savedServer, sshCredentials)
            } catch (e: Exception) {
                LLog.e(
                    "AppLifecycleController",
                    "manual reconnect failed",
                    e,
                    fields = mapOf("serverId" to serverId),
                )
            }
            appModel.refreshSnapshot()
            return
        }

        if (currentServer != null) {
            appModel.sshSessionStore.close(serverId)
            runCatching { appModel.serverBridge.disconnectServer(serverId) }
            try {
                appModel.serverBridge.connectRemoteServer(
                    serverId = currentServer.serverId,
                    displayName = currentServer.displayName,
                    host = currentServer.host,
                    port = currentServer.port,
                )
            } catch (e: Exception) {
                LLog.e(
                    "AppLifecycleController",
                    "manual reconnect fallback failed",
                    e,
                    fields = mapOf("serverId" to serverId),
                )
            }
            appModel.refreshSnapshot()
        }
    }

    private suspend fun reconnectSshServer(
        appModel: AppModel,
        server: SavedServer,
        credential: SavedSshCredential,
    ) {
        LLog.t(
            "AppLifecycleController",
            "reconnecting saved SSH server",
            fields = mapOf(
                "serverId" to server.id,
                "host" to server.hostname,
                "sshPort" to server.resolvedSshPort,
                "authMethod" to credential.method.name,
                "os" to server.os,
            ),
        )
        val ipcSocketPathOverride = ExperimentalFeatures.ipcSocketPathOverride()
        when (credential.method) {
            SshAuthMethod.PASSWORD -> {
                appModel.ssh.sshStartRemoteServerConnect(
                    serverId = server.id,
                    displayName = server.name,
                    host = server.hostname,
                    port = server.resolvedSshPort.toUShort(),
                    username = credential.username,
                    password = credential.password,
                    privateKeyPem = null,
                    passphrase = null,
                    acceptUnknownHost = true,
                    workingDir = null,
                    ipcSocketPathOverride = ipcSocketPathOverride,
                )
            }
            SshAuthMethod.KEY -> {
                appModel.ssh.sshStartRemoteServerConnect(
                    serverId = server.id,
                    displayName = server.name,
                    host = server.hostname,
                    port = server.resolvedSshPort.toUShort(),
                    username = credential.username,
                    password = null,
                    privateKeyPem = credential.privateKey,
                    passphrase = credential.passphrase,
                    acceptUnknownHost = true,
                    workingDir = null,
                    ipcSocketPathOverride = ipcSocketPathOverride,
                )
            }
        }
    }

    private suspend fun reconnectSavedServer(
        appModel: AppModel,
        server: SavedServer,
        sshCredentials: SshCredentialStore,
    ) {
        val directCodexPort = server.directCodexPort
        when {
            server.websocketURL != null -> {
                appModel.serverBridge.connectRemoteUrlServer(
                    serverId = server.id,
                    displayName = server.name,
                    websocketUrl = server.websocketURL,
                )
            }
            server.resolvedPreferredConnectionMode == "ssh" -> {
                val credential = sshCredentials.load(server.hostname, server.resolvedSshPort) ?: return
                reconnectSshServer(appModel, server, credential)
            }
            directCodexPort != null -> {
                appModel.serverBridge.connectRemoteServer(
                    serverId = server.id,
                    displayName = server.name,
                    host = server.hostname,
                    port = directCodexPort.toUShort(),
                )
            }
            else -> {
                LLog.t(
                    "AppLifecycleController",
                    "skipping reconnect; no valid saved transport",
                    fields = mapOf("serverId" to server.id),
                )
            }
        }
    }

    private suspend fun ensureLocalServerConnected(appModel: AppModel) {
        val currentLocal = appModel.store.snapshot()
            .servers
            .firstOrNull { it.isLocal && it.health != AppServerHealth.DISCONNECTED }
        if (currentLocal != null) {
            return
        }

        try {
            appModel.serverBridge.connectLocalServer(
                serverId = "local",
                displayName = "This Device",
                host = "127.0.0.1",
                port = 0u,
            )
            appModel.restoreStoredLocalChatGptAuth("local")
            appModel.refreshSessions(listOf("local"))
            appModel.refreshSnapshot()
            LLog.i("AppLifecycleController", "Local in-process server connected")
        } catch (e: Exception) {
            LLog.w(
                "AppLifecycleController",
                "local server reconnect failed",
                fields = mapOf("error" to e.message),
            )
        }
    }

    private suspend fun reconnectActiveSshServers(
        context: Context,
        appModel: AppModel,
        serverIds: Set<String>,
    ) {
        if (serverIds.isEmpty()) {
            return
        }

        val sshCredentials = SshCredentialStore(context)
        val savedServers = SavedServerStore.remembered(context)
            .filter { it.resolvedPreferredConnectionMode == "ssh" && it.id in serverIds }

        coroutineScope {
            savedServers.map { server ->
                async {
                    val credential = sshCredentials.load(server.hostname, server.resolvedSshPort) ?: return@async
                    try {
                        reconnectSshServer(appModel, server, credential)
                    } catch (e: Exception) {
                        LLog.w(
                            "AppLifecycleController",
                            "active SSH server reconnect failed",
                            fields = mapOf("serverId" to server.id, "error" to e.message),
                        )
                    }
                }
            }.awaitAll()
        }

        appModel.refreshSnapshot()
    }

    private suspend fun probeActiveRemoteServers(appModel: AppModel) {
        val serverIds = appModel.store.snapshot()
            .servers
            .filter { !it.isLocal && it.health == AppServerHealth.CONNECTED }
            .map { it.serverId }
            .distinct()

        if (serverIds.isEmpty()) {
            return
        }

        for (serverId in serverIds) {
            try {
                appModel.client.refreshAccount(
                    serverId,
                    AppRefreshAccountRequest(refreshToken = false),
                )
            } catch (e: Exception) {
                LLog.w(
                    "AppLifecycleController",
                    "active remote server probe failed",
                    fields = mapOf("serverId" to serverId, "error" to e.message),
                )
            }
        }

        appModel.refreshSnapshot()
    }

    /**
     * Called when the app enters the foreground.
     */
    suspend fun onResume(context: Context, appModel: AppModel) {
        val preResumeActiveSshServerIds = appModel.store.snapshot()
            .servers
            .filter { !it.isLocal && it.health != AppServerHealth.DISCONNECTED }
            .mapTo(mutableSetOf()) { it.serverId }
        ensureLocalServerConnected(appModel)
        reconnectSavedServers(context, appModel)
        reconnectActiveSshServers(context, appModel, preResumeActiveSshServerIds)
        probeActiveRemoteServers(appModel)
        backgroundedTurnKeys.clear()
    }

    /**
     * Called when the app goes to background.
     * Tracks active turns for notification on completion.
     */
    fun onPause(appModel: AppModel) {
        backgroundedTurnKeys.clear()
        val snap = appModel.snapshot.value ?: return
        for (thread in snap.threads) {
            if (thread.activeTurnId != null) {
                backgroundedTurnKeys.add(thread.key)
            }
        }
    }

    /**
     * Returns the set of threads that were active when the app was backgrounded.
     * Used by foreground service / push handler to know what to track.
     */
    fun getBackgroundedTurnKeys(): Set<ThreadKey> = backgroundedTurnKeys.toSet()
}

File: apps/ios/Sources/Litter/Models/AppLifecycleController.swift
import Foundation
import Observation
import UIKit
import UserNotifications

@MainActor
final class AppLifecycleController {
    static let notificationServerIdKey = "litter.notification.serverId"
    static let notificationThreadIdKey = "litter.notification.threadId"

    struct BackgroundTurnReconciliation {
        let remainingKeys: Set<ThreadKey>
        let activeThreads: [AppThreadSnapshot]
        let completedNotificationThread: AppThreadSnapshot?
    }

    private let pushProxy = PushProxyClient()
    private var pushProxyRegistrationId: String?
    private var devicePushToken: Data?
    private var backgroundedTurnKeys: Set<ThreadKey> = []
    private var backgroundTaskID: UIBackgroundTaskIdentifier = .invalid
    private var bgWakeCount: Int = 0
    private var notificationPermissionRequested = false
    private var hasRecoveredCurrentForegroundSession = false
    private var hasEnteredBackgroundSinceLaunch = false
    private var foregroundRecoveryTask: Task<Void, Never>?
    private var foregroundRecoveryID: UUID?

    func setDevicePushToken(_ token: Data) {
        devicePushToken = token
    }

    func reconnectSavedServers(appModel: AppModel) async {
        let plans = SavedServerStore.rememberedServers().compactMap {
            reconnectPlan(for: $0, appModel: appModel)
        }

        let tasks = plans.map { plan in
            Task { @MainActor [weak self] in
                guard let self else { return }
                await self.runReconnectPlan(plan, appModel: appModel)
            }
        }

        for task in tasks {
            await task.value
        }

        await appModel.refreshSnapshot()
    }

    func reconnectServer(serverId: String, appModel: AppModel) async {
        let snapshotServer = appModel.snapshot?.serverSnapshot(for: serverId)
        if snapshotServer?.isLocal == true || serverId == "local" {
            try? await appModel.restartLocalServer()
            return
        }

        if let savedServer = SavedServerStore.load().first(where: { $0.id == serverId }),
           let plan = reconnectPlan(
            for: savedServer,
            appModel: appModel,
            skipIfAlreadyConnected: false
           ) {
            await SshSessionStore.shared.close(serverId: serverId, ssh: appModel.ssh)
            appModel.serverBridge.disconnectServer(serverId: serverId)
            await runReconnectPlan(plan, appModel: appModel)
            await appModel.refreshSnapshot()
            return
        }

        await SshSessionStore.shared.close(serverId: serverId, ssh: appModel.ssh)
        appModel.serverBridge.disconnectServer(serverId: serverId)
        if let snapshotServer {
            _ = try? await appModel.serverBridge.connectRemoteServer(
                serverId: snapshotServer.serverId,
                displayName: snapshotServer.displayName,
                host: snapshotServer.host,
                port: snapshotServer.port
            )
        }
        await appModel.refreshSnapshot()
    }

    private func reconnectSSHServer(
        appModel: AppModel,
        serverId: String,
        displayName: String,
        host: String,
        port: UInt16,
        credentials: SSHCredentials
    ) async throws {
        let authMethod: String = switch credentials {
        case .password:
            "password"
        case .key:
            "private_key"
        }
        LLog.trace(
            "lifecycle",
            "reconnecting saved SSH server",
            fields: [
                "serverId": serverId,
                "host": host,
                "sshPort": Int(port),
                "authMethod": authMethod
            ]
        )
        let ipcSocketPathOverride = ExperimentalFeatures.shared.ipcSocketPathOverride()
        switch credentials {
        case .password(let username, let password):
            _ = try await appModel.ssh.sshStartRemoteServerConnect(
                serverId: serverId,
                displayName: displayName,
                host: host,
                port: port,
                username: username,
                password: password,
                privateKeyPem: nil,
                passphrase: nil,
                acceptUnknownHost: true,
                workingDir: nil,
                ipcSocketPathOverride: ipcSocketPathOverride
            )
        case .key(let username, let privateKey, let passphrase):
            _ = try await appModel.ssh.sshStartRemoteServerConnect(
                serverId: serverId,
                displayName: displayName,
                host: host,
                port: port,
                username: username,
                password: nil,
                privateKeyPem: privateKey,
                passphrase: passphrase,
                acceptUnknownHost: true,
                workingDir: nil,
                ipcSocketPathOverride: ipcSocketPathOverride
            )
        }
    }

    func appDidEnterBackground(
        snapshot: AppSnapshotRecord?,
        hasActiveVoiceSession: Bool,
        liveActivities: TurnLiveActivityController
    ) {
        hasEnteredBackgroundSinceLaunch = true
        hasRecoveredCurrentForegroundSession = false
        foregroundRecoveryTask?.cancel()
        foregroundRecoveryTask = nil
        foregroundRecoveryID = nil
        guard !hasActiveVoiceSession else { return }
        let activeThreads = snapshot?.threadsWithTrackedTurns ?? []
        guard !activeThreads.isEmpty else { return }

        backgroundedTurnKeys = Set(activeThreads.map(\.key))
        bgWakeCount = 0
        liveActivities.sync(snapshot)
        registerPushProxy()

        let bgID = UIApplication.shared.beginBackgroundTask { [weak self] in
            guard let self else { return }
            let expiredID = self.backgroundTaskID
            self.backgroundTaskID = .invalid
            UIApplication.shared.endBackgroundTask(expiredID)
        }
        backgroundTaskID = bgID
    }

    func appDidBecomeActive(
        appModel: AppModel,
        hasActiveVoiceSession: Bool,
        liveActivities: TurnLiveActivityController
    ) {
        deregisterPushProxy()
        endBackgroundTaskIfNeeded()
        guard !hasActiveVoiceSession else { return }
        guard !hasRecoveredCurrentForegroundSession else { return }
        hasRecoveredCurrentForegroundSession = true
        let needsInitialReconnect = !hasEnteredBackgroundSinceLaunch
        let preResumeActiveSSHServerIDs = Set((appModel.snapshot?.servers ?? [])
            .filter { !$0.isLocal && $0.health != .disconnected }
            .map(\.serverId))
        let currentSnapshot = appModel.snapshot
        let backgroundedKeys = backgroundedTurnKeys
        backgroundedTurnKeys.removeAll()
        var keysToRefresh = Set(currentSnapshot?.threads.compactMap { thread in
            currentSnapshot?.threadHasTrackedTurn(for: thread.key) == true ? thread.key : nil
        } ?? [])
        if let activeKey = currentSnapshot?.activeThread {
            keysToRefresh.insert(activeKey)
        }

        foregroundRecoveryTask?.cancel()
        let recoveryID = UUID()
        foregroundRecoveryID = recoveryID

        foregroundRecoveryTask = Task { [weak self] in
            guard let self else { return }
            defer {
                if self.foregroundRecoveryID == recoveryID {
                    self.foregroundRecoveryTask = nil
                    self.foregroundRecoveryID = nil
                }
            }

            await self.performForegroundRecovery(
                appModel: appModel,
                liveActivities: liveActivities,
                needsInitialReconnect: needsInitialReconnect,
                reconnectActiveServerIDs: preResumeActiveSSHServerIDs,
                backgroundedKeys: backgroundedKeys,
                keysToRefresh: keysToRefresh
            )
        }
    }

    func handleBackgroundPush(
        appModel: AppModel,
        liveActivities: TurnLiveActivityController
    ) async {
        bgWakeCount += 1
        let keys = backgroundedTurnKeys
        guard !keys.isEmpty else { return }

        await reconnectSavedServers(appModel: appModel)
        await refreshTrackedThreads(appModel: appModel, keys: Array(keys))
        await appModel.refreshSnapshot()

        guard let snapshot = appModel.snapshot else { return }
        let reconciliation = reconcileBackgroundedTurns(snapshot: snapshot, trackedKeys: keys)
        backgroundedTurnKeys = reconciliation.remainingKeys

        for thread in reconciliation.activeThreads {
            liveActivities.updateBackgroundWake(for: thread, pushCount: bgWakeCount)
        }

        if let thread = reconciliation.completedNotificationThread {
            liveActivities.endCurrent(phase: .completed, snapshot: snapshot)
            postLocalNotificationIfNeeded(
                model: thread.resolvedModel,
                threadPreview: thread.resolvedPreview,
                threadKey: thread.key
            )
        }

        if backgroundedTurnKeys.isEmpty {
            deregisterPushProxy()
        }
    }

    func requestNotificationPermissionIfNeeded() {
        guard !notificationPermissionRequested else { return }
        notificationPermissionRequested = true
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { _, _ in }
    }

    private func reconnectActiveSSHServers(
        appModel: AppModel,
        serverIDs: Set<String>
    ) async {
        guard !serverIDs.isEmpty else { return }

        let plans = SavedServerStore.rememberedServers().compactMap { savedServer -> SavedReconnectPlan? in
            guard savedServer.preferredConnectionMode == .ssh,
                  serverIDs.contains(savedServer.id) else {
                return nil
            }
            let server = savedServer.toDiscoveredServer()
            guard let credential = try? SSHCredentialStore.shared.load(
                host: server.hostname,
                port: Int(server.resolvedSSHPort)
            ) else {
                return nil
            }
            return .ssh(
                serverId: server.id,
                displayName: server.name,
                host: server.hostname,
                port: server.resolvedSSHPort,
                credentials: credential.toConnectionCredential()
            )
        }

        let tasks = plans.map { plan in
            Task { @MainActor [weak self] in
                guard let self else { return }
                await self.runReconnectPlan(plan, appModel: appModel)
            }
        }

        for task in tasks {
            await task.value
        }

        await appModel.refreshSnapshot()
    }

    func reconcileBackgroundedTurns(
        snapshot: AppSnapshotRecord,
        trackedKeys: Set<ThreadKey>
    ) -> BackgroundTurnReconciliation {
        var remainingKeys: Set<ThreadKey> = []
        var activeThreads: [AppThreadSnapshot] = []
        var completedThreads: [AppThreadSnapshot] = []

        for key in trackedKeys {
            guard let thread = snapshot.threadSnapshot(for: key) else {
                // Keep tracking until we can observe a definitive thread state again.
                remainingKeys.insert(key)
                continue
            }

            if snapshot.threadHasTrackedTurn(for: key) {
                remainingKeys.insert(key)
                activeThreads.append(thread)
            } else {
                completedThreads.append(thread)
            }
        }

        let completedNotificationThread: AppThreadSnapshot?
        if remainingKeys.isEmpty {
            completedNotificationThread = completedThreads.first(where: {
                $0.info.parentThreadId == nil
            }) ?? completedThreads.first
        } else {
            completedNotificationThread = nil
        }

        return BackgroundTurnReconciliation(
            remainingKeys: remainingKeys,
            activeThreads: activeThreads,
            completedNotificationThread: completedNotificationThread
        )
    }

    private func reconnectPlan(
        for savedServer: SavedServer,
        appModel: AppModel,
        skipIfAlreadyConnected: Bool = true
    ) -> SavedReconnectPlan? {
        let server = savedServer.toDiscoveredServer()
        if skipIfAlreadyConnected,
           let snapshot = appModel.snapshot?.serverSnapshot(for: server.id),
           snapshot.health != .disconnected {
            return nil
        }

        do {
            if savedServer.preferredConnectionMode == .ssh {
                guard let credential = try SSHCredentialStore.shared.load(
                    host: server.hostname,
                    port: Int(server.resolvedSSHPort)
                ) else {
                    return nil
                }
                return .ssh(
                    serverId: server.id,
                    displayName: server.name,
                    host: server.hostname,
                    port: server.resolvedSSHPort,
                    credentials: credential.toConnectionCredential()
                )
            } else if let target = server.connectionTarget {
                switch target {
                case .local:
                    return .local(
                        serverId: server.id,
                        displayName: server.name,
                        restoreLocalAuth: true
                    )
                case .remote(let host, let port):
                    return .remote(
                        serverId: server.id,
                        displayName: server.name,
                        host: host,
                        port: port
                    )
                case .remoteURL(let url):
                    return .remoteURL(
                        serverId: server.id,
                        displayName: server.name,
                        websocketUrl: url.absoluteString
                    )
                case .sshThenRemote(let host, let credentials):
                    return .ssh(
                        serverId: server.id,
                        displayName: server.name,
                        host: host,
                        port: server.resolvedSSHPort,
                        credentials: credentials
                    )
                }
            } else if savedServer.preferredConnectionMode == nil,
                      let credential = try SSHCredentialStore.shared.load(
                host: server.hostname,
                port: Int(server.resolvedSSHPort)
            ) {
                return .ssh(
                    serverId: server.id,
                    displayName: server.name,
                    host: server.hostname,
                    port: server.resolvedSSHPort,
                    credentials: credential.toConnectionCredential()
                )
            }
        } catch {
            return nil
        }

        return nil
    }

    private func runReconnectPlan(
        _ plan: SavedReconnectPlan,
        appModel: AppModel
    ) async {
        do {
            switch plan {
            case .ssh(let serverId, let displayName, let host, let port, let credentials):
                try await reconnectSSHServer(
                    appModel: appModel,
                    serverId: serverId,
                    displayName: displayName,
                    host: host,
                    port: port,
                    credentials: credentials
                )
            case .local(let serverId, let displayName, let restoreLocalAuth):
                _ = try await appModel.serverBridge.connectLocalServer(
                    serverId: serverId,
                    displayName: displayName,
                    host: "127.0.0.1",
                    port: 0
                )
                if restoreLocalAuth {
                    await appModel.restoreStoredLocalChatGPTAuth(serverId: serverId)
                }
            case .remote(let serverId, let displayName, let host, let port):
                _ = try await appModel.serverBridge.connectRemoteServer(
                    serverId: serverId,
                    displayName: displayName,
                    host: host,
                    port: port
                )
            case .remoteURL(let serverId, let displayName, let websocketUrl):
                _ = try await appModel.serverBridge.connectRemoteUrlServer(
                    serverId: serverId,
                    displayName: displayName,
                    websocketUrl: websocketUrl
                )
            }
        } catch {}
    }

    private enum SavedReconnectPlan {
        case ssh(
            serverId: String,
            displayName: String,
            host: String,
            port: UInt16,
            credentials: SSHCredentials
        )
        case local(
            serverId: String,
            displayName: String,
            restoreLocalAuth: Bool
        )
        case remote(
            serverId: String,
            displayName: String,
            host: String,
            port: UInt16
        )
        case remoteURL(
            serverId: String,
            displayName: String,
            websocketUrl: String
        )
    }

    private func performForegroundRecovery(
        appModel: AppModel,
        liveActivities: TurnLiveActivityController,
        needsInitialReconnect: Bool,
        reconnectActiveServerIDs: Set<String>,
        backgroundedKeys: Set<ThreadKey>,
        keysToRefresh: Set<ThreadKey>
    ) async {
        if needsInitialReconnect {
            await reconnectSavedServers(appModel: appModel)
            guard !Task.isCancelled else { return }
        }

        let serverIDsToReconnect: Set<String>
        if needsInitialReconnect {
            serverIDsToReconnect = reconnectActiveServerIDs
        } else {
            let focusedServerIDs = Set(backgroundedKeys.map(\.serverId))
                .union(keysToRefresh.map(\.serverId))
            serverIDsToReconnect = reconnectActiveServerIDs.intersection(focusedServerIDs)
        }

        if !serverIDsToReconnect.isEmpty {
            await reconnectActiveSSHServers(
                appModel: appModel,
                serverIDs: serverIDsToReconnect
            )
            guard !Task.isCancelled else { return }
        }

        if !keysToRefresh.isEmpty {
            await refreshTrackedThreads(appModel: appModel, keys: Array(keysToRefresh))
            guard !Task.isCancelled else { return }
        }

        await appModel.refreshSnapshot()
        liveActivities.sync(appModel.snapshot)
    }

    private func refreshTrackedThreads(appModel: AppModel, keys: [ThreadKey]) async {
        let serverIds = Set(keys.map(\.serverId))
        for serverId in serverIds {
            _ = try? await appModel.client.listThreads(
                serverId: serverId,
                params: AppListThreadsRequest(
                    cursor: nil,
                    limit: nil,
                    archived: nil,
                    cwd: nil,
                    searchTerm: nil
                )
            )
        }

        let snapshot = appModel.snapshot
        for key in keys {
            let existing = snapshot?.threadSnapshot(for: key)
            let cwd = existing?.info.cwd?.trimmingCharacters(in: .whitespacesAndNewlines)
            let config = AppThreadLaunchConfig(
                model: existing?.resolvedModel,
                approvalPolicy: nil,
                sandbox: nil,
                developerInstructions: nil,
                persistExtendedHistory: true
            )
            _ = try? await appModel.client.resumeThread(
                serverId: key.serverId,
                params: config.threadResumeRequest(
                    threadId: key.threadId,
                    cwdOverride: cwd?.isEmpty == false ? cwd : nil
                )
            )
        }
    }

    private func registerPushProxy() {
        guard let tokenData = devicePushToken else { return }
        guard pushProxyRegistrationId == nil else { return }
        let token = tokenData.map { String(format: "%02x", $0) }.joined()
        Task {
            do {
                let regId = try await pushProxy.register(pushToken: token, interval: 30, ttl: 7200)
                await MainActor.run {
                    self.pushProxyRegistrationId = regId
                }
            } catch {}
        }
    }

    private func deregisterPushProxy() {
        guard let regId = pushProxyRegistrationId else { return }
        pushProxyRegistrationId = nil
        Task {
            try? await pushProxy.deregister(registrationId: regId)
        }
    }

    private func endBackgroundTaskIfNeeded() {
        guard backgroundTaskID != .invalid else { return }
        UIApplication.shared.endBackgroundTask(backgroundTaskID)
        backgroundTaskID = .invalid
    }

    static func notificationThreadKey(from userInfo: [AnyHashable: Any]) -> ThreadKey? {
        guard let serverId = userInfo[notificationServerIdKey] as? String,
              let threadId = userInfo[notificationThreadIdKey] as? String else {
            return nil
        }

        let trimmedServerId = serverId.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedThreadId = threadId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedServerId.isEmpty, !trimmedThreadId.isEmpty else { return nil }

        return ThreadKey(serverId: trimmedServerId, threadId: trimmedThreadId)
    }

    private func postLocalNotificationIfNeeded(
        model: String,
        threadPreview: String?,
        threadKey: ThreadKey
    ) {
        guard UIApplication.shared.applicationState != .active else { return }
        let content = UNMutableNotificationContent()
        content.title = "Turn completed"
        var bodyParts: [String] = []
        if let preview = threadPreview, !preview.isEmpty { bodyParts.append(preview) }
        if !model.isEmpty { bodyParts.append(model) }
        content.body = bodyParts.joined(separator: " - ")
        content.sound = .default
        content.userInfo = [
            Self.notificationServerIdKey: threadKey.serverId,
            Self.notificationThreadIdKey: threadKey.threadId
        ]
        let request = UNNotificationRequest(
            identifier: UUID().uuidString,
            content: content,
            trigger: nil
        )
        UNUserNotificationCenter.current().add(request)
    }
}
